### PR TITLE
#22: Improved logging

### LINKED ISSUE #22: Improve logging
## Background

Currently, everything is logged to the default logger. Also the log-level is configured independently of the log channel name. 
Most of the progress is logged to log level info to the default logger. 

Now, if we do not set `--log-level info`, we can't see any progress of the ongoing tasks.
But, if we set `--log-level info`, the progress and the AWS boto3 logs are getting mixed up.

As a user I only want to see the relevant logs showing the progress of the export vm tasks....but not any boto3 internal logs.

## Acceptance Criteria

1. Move the logs relevant for the user to a new separate logger
2. Change log configuration so that info for the new logger is default-wise on.

### PR DESCRIPTION
**Pull request by tomuben**
fixes #22 

Use [RichHandler](https://rich.readthedocs.io/en/stable/reference/logging.html#logging) and defined separate loggers for each module.
image

### CODE REVIEW

**Review by Nicoretti — COMMENTED** [review]
General note on the log messages: In general I am also a huge fan of 'f-Strings' when it comes to log messages the build in printf like formatting makes sense though. 

TL;DR:
* Performance
* Delaying/Evaluating the expressions to the point when (if) they are outputted

If you want to go a bit more into the details [this blog article](https://dev.to/izabelakowal/what-is-the-best-string-formatting-technique-for-logging-in-python-d1d) may be a good start.

File: exasol_script_languages_developer_sandbox/lib/aws_access/aws_access.py [hunk 1]
```diff
@@ -14,13 +13,15 @@
...
 from exasol_script_languages_developer_sandbox.lib.aws_access.stack_resource import StackResource
+from exasol_script_languages_developer_sandbox.lib.logging import get_status_logger, LogType
 from exasol_script_languages_developer_sandbox.lib.tags import create_default_asset_tag
 from exasol_script_languages_developer_sandbox.lib.export_vm.vm_disk_image_format import VmDiskImageFormat
 
 
 class AwsAccess(object):
     def __init__(self, aws_profile: Optional[str]):
         self._aws_profile = aws_profile
+        self._logger = get_status_logger(LogType.AWS_ACCESS)
```
Comment (tkilias) [inline exasol_script_languages_developer_sandbox/lib/aws_access/aws_access.py]: Does each object need its own logger

File: exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
-import logging
 from pathlib import Path
 import importlib_resources as ir
 import exasol_script_languages_developer_sandbox.runtime.ansible
+from exasol_script_languages_developer_sandbox.lib.logging import get_status_logger, LogType
+
+LOG = get_status_logger(LogType.ANSIBLE)
```
Comment (tkilias) [inline exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py:6]: A common pattern for creating logger is to use the module name , is there a specific reason you are using specific categories

File: exasol_script_languages_developer_sandbox/lib/logging.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    SETUP = "setup"
+    EXPORT = "export"
+    VM_BUCKET = "vm_bucket"
+    CI_CODEBUILD = "ci_codebuild"
+    AWS_ACCESS = "aws_access"
+    ANSIBLE = "ansible"
+
+
+def get_status_logger(log_type: LogType) -> logging.Logger:
+    return logging.getLogger(f"eslds-{log_type.value}")
```
Comment (tkilias) [inline exasol_script_languages_developer_sandbox/lib/logging.py:19]: What is eslds

**Review by tkilias — CHANGES_REQUESTED** [review]

**Review by Nicoretti — COMMENTED** [review]
I have seen you have quite a lot of `debug` log messages which seem to repeat a common pattern. The common pattern seems to be to log that function `x` is executed you could simplify this e.g. by writing a decorator for this.

Example:
```python
from functools import wraps


def log_execution(logger, level):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            logger(f"{level}: Entering {func.__name__}")
            result = func(*args, **kwargs)
            logger(f"{level}: Leaving {func.__name__}")
            return result

        return wrapper

    return decorator


@log_execution(print, "DEBUG")
def some_func(arg):
    print(f"doing work, arg = {arg}")


@log_execution(print, "INFO")
def some_other_func(foo, bar=1):
    print(f"doing even more work, foo = {foo}, bar = {bar}")


if __name__ == "__main__":
    some_func(1)
    some_other_func(2)
```

**Comment by tkilias** [comment]
> General note on the log messages: In general I am also a huge fan of 'f-Strings' when it comes to log messages the build in printf like formatting makes sense though.
> 
> TL;DR:
> 
> * Performance
> * Delaying/Evaluating the expressions to the point when (if) they are outputted
> 
> If you want to go a bit more into the details [this blog article](https://dev.to/izabelakowal/what-is-the-best-string-formatting-technique-for-logging-in-python-d1d) may be a good start.

@Nicoretti I think, performance doesn't matter here. The main steps take minutes to hours. A few more milliseconds we can ignore

**Comment by Nicoretti** [comment]
> > General note on the log messages: In general I am also a huge fan of 'f-Strings' when it comes to log messages the build in printf like formatting makes sense though.
> > TL;DR:
> > 
> > * Performance
> > * Delaying/Evaluating the expressions to the point when (if) they are outputted
> > 
> > If you want to go a bit more into the details [this blog article](https://dev.to/izabelakowal/what-is-the-best-string-formatting-technique-for-logging-in-python-d1d) may be a good start.
> 
> @Nicoretti I think, performance doesn't matter here. The main steps take minutes to hours. A few more milliseconds we can ignore

I get your point, still I personally think it makes sense to do this distinction always, so one will develop a habit of doing the logging differently so it won't get missed where/if it counts.

File: exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py (hunk 2 above)
Comment (tomuben) [reply exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py:6]: The module name would be actually the file name, however this is already automatically printed by using the RichHandler.
I attached a screenshot for better understanding.
I create a logger per each submodule, which helps to communicate to the user which step the tool is currently doing.

File: exasol_script_languages_developer_sandbox/lib/logging.py (hunk 3 above)
Comment (tomuben) [reply exasol_script_languages_developer_sandbox/lib/logging.py:19]: **E**xasol-**S**cript-**L**anguages-**D**eveloper-**S**andbox

File: exasol_script_languages_developer_sandbox/lib/aws_access/aws_access.py (hunk 1 above)
Comment (tomuben) [reply exasol_script_languages_developer_sandbox/lib/aws_access/aws_access.py]: thinking well, it's better to move the logger variable to the module as it's done in the other files.

Comment (tomuben) [reply exasol_script_languages_developer_sandbox/lib/aws_access/aws_access.py]: done

**Comment by tomuben** [comment]
> I have seen you have quite a lot of `debug` log messages which seem to repeat a common pattern. The common pattern seems to be to log that function `x` is executed you could simplify this e.g. by writing a decorator for this.
> 
> Example:
> 
> ```python
> from functools import wraps
> 
> 
> def log_execution(logger, level):
>     def decorator(func):
>         @wraps(func)
>         def wrapper(*args, **kwargs):
>             logger(f"{level}: Entering {func.__name__}")
>             result = func(*args, **kwargs)
>             logger(f"{level}: Leaving {func.__name__}")
>             return result
> 
>         return wrapper
> 
>     return decorator
> 
> 
> @log_execution(print, "DEBUG")
> def some_func(arg):
>     print(f"doing work, arg = {arg}")
> 
> 
> @log_execution(print, "INFO")
> def some_other_func(foo, bar=1):
>     print(f"doing even more work, foo = {foo}, bar = {bar}")
> 
> 
> if __name__ == "__main__":
>     some_func(1)
>     some_other_func(2)
> ```

Added your suggestion

**Comment by tomuben** [comment]
> > > General note on the log messages: In general I am also a huge fan of 'f-Strings' when it comes to log messages the build in printf like formatting makes sense though.
> > > TL;DR:
> > > 
> > > * Performance
> > > * Delaying/Evaluating the expressions to the point when (if) they are outputted
> > > 
> > > If you want to go a bit more into the details [this blog article](https://dev.to/izabelakowal/what-is-the-best-string-formatting-technique-for-logging-in-python-d1d) may be a good start.
> > 
> > 
> > @Nicoretti I think, performance doesn't matter here. The main steps take minutes to hours. A few more milliseconds we can ignore
> 
> I get your point, still I personally think it makes sense to do this distinction always, so one will develop a habit of doing the logging differently so it won't get missed where/if it counts.

Maybe that's an interesting point for our coding guidelines.
As Torsten said: Performance is not criticial for this project, so I think it's not worth the effort to change all the logs now. 
But I will keep it in mind for future changes.

File: exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py (hunk 2 above)
Comment (tomuben) [reply exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py:6]: Also, I need to have a list of all custom loggers for which I want to set the log-level to (at least) info.

**Review by tkilias — DISMISSED** [review]

**Review by Nicoretti — APPROVED** [review]
